File: portfolio-backend/app/core/validators.py
"""
Comprehensive Input Validation & Sanitization Module

This module provides enterprise-grade input validation and sanitization
to prevent common security vulnerabilities:
- SQL Injection
- XSS (Cross-Site Scripting)
- Path Traversal
- Command Injection
- LDAP Injection
- XML External Entity (XXE)

Usage:
    from app.core.validators import InputValidator, sanitize_html, validate_url
    
    validator = InputValidator()
    
    # SQL injection check
    is_safe, msg = validator.validate_sql_safe("user input")
    
    # XSS sanitization
    clean_text = sanitize_html(user_html)
    
    # URL validation
    is_valid, url = validate_url("https://example.com")
"""

import re
import html
import unicodedata
from typing import Tuple, Optional, List, Dict, Any
from urllib.parse import urlparse, quote, unquote
from pathlib import Path
import json

# Dangerous patterns for various injection attacks
SQL_INJECTION_PATTERNS = [
    r"(\b(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE|UNION|SCRIPT|JAVASCRIPT)\b)",
    r"(--|\#|\/\*|\*\/)",  # SQL comments
    r"(\bOR\b.*=.*)",  # OR clauses
    r"(;|\||&|`|\$\(|\$\{)",  # Command separators
    r"(xp_|sp_)",  # SQL Server stored procedures
    r"(\bAND\b.*=.*)",  # AND clauses in suspicious context
]

XSS_PATTERNS = [
    r"<script[^>]*>.*?</script>",  # Script tags
    r"javascript:",  # JavaScript protocol
    r"on\w+\s*=",  # Event handlers (onclick, onerror, etc.)
    r"<iframe[^>]*>",  # IFrames
    r"<embed[^>]*>",  # Embed tags
    r"<object[^>]*>",  # Object tags
    r"eval\s*\(",  # eval() calls
    r"expression\s*\(",  # CSS expressions
    r"vbscript:",  # VBScript protocol
    r"data:text/html",  # Data URIs with HTML
]

PATH_TRAVERSAL_PATTERNS = [
    r"\.\./",  # Directory traversal
    r"\.\.",  # Parent directory
    r"~",  # Home directory
    r"/etc/",  # System directories
    r"/proc/",
    r"/sys/",
    r"\\\\",  # UNC paths
    r"file://",  # File protocol
]

COMMAND_INJECTION_PATTERNS = [
    r"[;&|`$]",  # Shell metacharacters
    r"\$\(.*\)",  # Command substitution
    r"`.*`",  # Backticks
    r"\|\|",  # OR operator
    r"&&",  # AND operator
]

LDAP_INJECTION_PATTERNS = [
    r"[*()\\]",  # LDAP special characters
    r"[\x00]",  # Null byte
]

# Allowed HTML tags and attributes for rich text
ALLOWED_HTML_TAGS = {
    'p', 'br', 'strong', 'em', 'u', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
    'blockquote', 'code', 'pre', 'hr', 'ul', 'ol', 'li', 'a', 'img',
    'table', 'thead', 'tbody', 'tr', 'th', 'td', 'span', 'div'
}

ALLOWED_HTML_ATTRIBUTES = {
    'a': {'href', 'title', 'target'},
    'img': {'src', 'alt', 'title', 'width', 'height'},
    'span': {'class'},
    'div': {'class'},
    'td': {'colspan', 'rowspan'},
    'th': {'colspan', 'rowspan'},
}

# Dangerous file extensions
DANGEROUS_EXTENSIONS = {
    'exe', 'bat', 'cmd', 'com', 'pif', 'scr', 'vbs', 'js', 'jar',
    'dll', 'sys', 'sh', 'bash', 'ps1', 'psm1', 'app', 'deb', 'rpm'
}


class InputValidator:
    """
    Comprehensive input validator for security checks.
    
    Validates inputs against various injection attack patterns
    and provides sanitization utilities.
    """
    
    def __init__(self, strict_mode: bool = True):
        """
        Initialize validator.
        
        Args:
            strict_mode: If True, applies stricter validation rules
        """
        self.strict_mode = strict_mode
    
    def validate_sql_safe(self, value: str) -> Tuple[bool, Optional[str]]:
        """
        Check if input is safe from SQL injection.
        
        Args:
            value: String to validate
            
        Returns:
            Tuple of (is_safe, error_message)
        """
        if not value:
            return True, None
        
        value_upper = value.upper()
        
        for pattern in SQL_INJECTION_PATTERNS:
            if re.search(pattern, value_upper, re.IGNORECASE):
                return False, f"Potential SQL injection detected: suspicious pattern found"
        
        return True, None
    
    def validate_xss_safe(self, value: str) -> Tuple[bool, Optional[str]]:
        """
        Check if input is safe from XSS attacks.
        
        Args:
            value: String to validate
            
        Returns:
            Tuple of (is_safe, error_message)
        """
        if not value:
            return True, None
        
        for pattern in XSS_PATTERNS:
            if re.search(pattern, value, re.IGNORECASE):
                return False, f"Potential XSS attack detected: suspicious pattern found"
        
        return True, None
    
    def validate_path_safe(self, value: str) -> Tuple[bool, Optional[str]]:
        """
        Check if path is safe from traversal attacks.
        
        Args:
            value: Path to validate
            
        Returns:
            Tuple of (is_safe, error_message)
        """
        if not value:
            return True, None
        
        for pattern in PATH_TRAVERSAL_PATTERNS:
            if re.search(pattern, value, re.IGNORECASE):
                return False, f"Potential path traversal detected: suspicious pattern found"
        
        # Additional check: resolve path and ensure it doesn't escape
        try:
            resolved_path = Path(value).resolve()
            # This is a basic check; in production, compare against allowed base paths
            if str(resolved_path).startswith('/etc') or str(resolved_path).startswith('/sys'):
                return False, "Access to system directories not allowed"
        except Exception:
            return False, "Invalid path format"
        
        return True, None
    
    def validate_command_safe(self, value: str) -> Tuple[bool, Optional[str]]:
        """
        Check if input is safe from command injection.
        
        Args:
            value: String to validate
            
        Returns:
            Tuple of (is_safe, error_message)
        """
        if not value:
            return True, None
        
        for pattern in COMMAND_INJECTION_PATTERNS:
            if re.search(pattern, value):
                return False, f"Potential command injection detected: suspicious pattern found"
        
        return True, None
    
    def validate_ldap_safe(self, value: str) -> Tuple[bool, Optional[str]]:
        """
        Check if input is safe for LDAP queries.
        
        Args:
            value: String to validate
            
        Returns:
            Tuple of (is_safe, error_message)
        """
        if not value:
            return True, None
        
        for pattern in LDAP_INJECTION_PATTERNS:
            if re.search(pattern, value):
                return False, f"Potential LDAP injection detected: suspicious character found"
        
        return True, None
    
    def validate_json_safe(self, value: str) -> Tuple[bool, Optional[str]]:
        """
        Check if JSON string is valid and safe.
        
        Args:
            value: JSON string to validate
            
        Returns:
            Tuple of (is_safe, error_message)
        """
        if not value:
            return True, None
        
        try:
            parsed = json.loads(value)
            # Check for excessively deep nesting (DoS prevention)
            if self._get_json_depth(parsed) > 10:
                return False, "JSON nesting too deep (max 10 levels)"
            return True, None
        except json.JSONDecodeError as e:
            return False, f"Invalid JSON: {str(e)}"
    
    def _get_json_depth(self, obj: Any, depth: int = 0) -> int:
        """Calculate maximum depth of JSON object."""
        if isinstance(obj, dict):
            return max([self._get_json_depth(v, depth + 1) for v in obj.values()], default=depth)
        elif isinstance(obj, list):
            return max([self._get_json_depth(item, depth + 1) for item in obj], default=depth)
        return depth
    
    def validate_filename(self, filename: str) -> Tuple[bool, Optional[str]]:
        """
        Validate filename for security issues.
        
        Args:
            filename: Filename to validate
            
        Returns:
            Tuple of (is_safe, error_message)
        """
        if not filename:
            return False, "Filename cannot be empty"
        
        # Check for path traversal
        if '..' in filename or '/' in filename or '\\' in filename:
            return False, "Filename cannot contain path separators or parent directory references"
        
        # Check for control characters
        if any(ord(char) < 32 for char in filename):
            return False, "Filename cannot contain control characters"
        
        # Check extension
        ext = filename.rsplit('.', 1)[-1].lower() if '.' in filename else ''
        if ext in DANGEROUS_EXTENSIONS:
            return False, f"File extension '.{ext}' is not allowed for security reasons"
        
        # Check length
        if len(filename) > 255:
            return False, "Filename too long (max 255 characters)"
        
        return True, None
    
    def validate_email(self, email: str) -> Tuple[bool, Optional[str]]:
        """
        Validate email address format.
        
        Args:
            email: Email address to validate
            
        Returns:
            Tuple of (is_valid, error_message)
        """
        if not email:
            return False, "Email cannot be empty"
        
        # Basic email regex (RFC 5322 simplified)
        email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        
        if not re.match(email_pattern, email):
            return False, "Invalid email format"
        
        # Additional checks
        if len(email) > 254:
            return False, "Email too long (max 254 characters)"
        
        local_part, domain = email.rsplit('@', 1)
        if len(local_part) > 64:
            return False, "Email local part too long (max 64 characters)"
        
        return True, None
    
    def validate_username(self, username: str) -> Tuple[bool, Optional[str]]:
        """
        Validate username format.
        
        Args:
            username: Username to validate
            
        Returns:
            Tuple of (is_valid, error_message)
        """
        if not username:
            return False, "Username cannot be empty"
        
        # Only allow alphanumeric, underscore, hyphen, period
        if not re.match(r'^[a-zA-Z0-9._-]+$', username):
            return False, "Username can only contain letters, numbers, and ._- characters"
        
        if len(username) < 3:
            return False, "Username must be at least 3 characters"
        
        if len(username) > 50:
            return False, "Username cannot exceed 50 characters"
        
        return True, None
    
    def sanitize_string(self, value: str, max_length: Optional[int] = None) -> str:
        """
        Sanitize a string by removing dangerous characters.
        
        Args:
            value: String to sanitize
            max_length: Maximum allowed length
            
        Returns:
            Sanitized string
        """
        if not value:
            return ""
        
        # Normalize unicode characters
        value = unicodedata.normalize('NFKC', value)
        
        # Remove control characters (except common whitespace)
        value = ''.join(char for char in value if ord(char) >= 32 or char in '\n\r\t')
        
        # Remove null bytes
        value = value.replace('\x00', '')
        
        # Trim whitespace
        value = value.strip()
        
        # Truncate if needed
        if max_length and len(value) > max_length:
            value = value[:max_length]
        
        return value
    
    def comprehensive_check(self, value: str, checks: Optional[List[str]] = None) -> Tuple[bool, List[str]]:
        """
        Run multiple validation checks on input.
        
        Args:
            value: String to validate
            checks: List of check names to run. If None, runs all checks.
                   Options: 'sql', 'xss', 'path', 'command', 'ldap'
        
        Returns:
            Tuple of (is_safe, list_of_errors)
        """
        if checks is None:
            checks = ['sql', 'xss', 'command']
        
        errors = []
        
        if 'sql' in checks:
            is_safe, msg = self.validate_sql_safe(value)
            if not is_safe:
                errors.append(msg)
        
        if 'xss' in checks:
            is_safe, msg = self.validate_xss_safe(value)
            if not is_safe:
                errors.append(msg)
        
        if 'path' in checks:
            is_safe, msg = self.validate_path_safe(value)
            if not is_safe:
                errors.append(msg)
        
        if 'command' in checks:
            is_safe, msg = self.validate_command_safe(value)
            if not is_safe:
                errors.append(msg)
        
        if 'ldap' in checks:
            is_safe, msg = self.validate_ldap_safe(value)
            if not is_safe:
                errors.append(msg)
        
        return len(errors) == 0, errors


def sanitize_html(html_content: str, allowed_tags: Optional[set] = None, 
                   allowed_attributes: Optional[Dict[str, set]] = None) -> str:
    """
    Sanitize HTML content by removing dangerous tags and attributes.
    
    This is a lightweight sanitizer. For full HTML sanitization,
    use app.utils.html_sanitizer module which uses bleach library.
    
    Args:
        html_content: HTML string to sanitize
        allowed_tags: Set of allowed HTML tags (default: escape all)
        allowed_attributes: Dict of allowed attributes per tag
        
    Returns:
        Sanitized HTML string
    """
    if not html_content:
        return ""
    
    # For security, default to escaping all HTML
    # Use app.utils.html_sanitizer for rich text that needs formatting
    return html.escape(html_content)


def sanitize_url(url: str) -> str:
    """
    Sanitize URL by encoding special characters.
    
    Args:
        url: URL to sanitize
        
    Returns:
        Sanitized URL
    """
    if not url:
        return ""
    
    # Remove control characters
    url = ''.join(char for char in url if ord(char) >= 32)
    
    # Basic URL encoding
    return quote(url, safe=':/?#[]@!$&\'()*+,;=')


def validate_url(url: str, allowed_schemes: Optional[List[str]] = None) -> Tuple[bool, Optional[str]]:
    """
    Validate URL format and scheme.
    
    Args:
        url: URL to validate
        allowed_schemes: List of allowed schemes (default: ['http', 'https'])
        
    Returns:
        Tuple of (is_valid, error_message or validated_url)
    """
    if not url:
        return False, "URL cannot be empty"
    
    if allowed_schemes is None:
        allowed_schemes = ['http', 'https']
    
    try:
        parsed = urlparse(url)
        
        # Check scheme
        if parsed.scheme not in allowed_schemes:
            return False, f"URL scheme must be one of: {', '.join(allowed_schemes)}"
        
        # Check for valid domain
        if not parsed.netloc:
            return False, "URL must have a valid domain"
        
        # Check for suspicious patterns
        if any(char in url for char in ['<', '>', '"', "'"]):
            return False, "URL contains invalid characters"
        
        # Check length
        if len(url) > 2048:
            return False, "URL too long (max 2048 characters)"
        
        return True, url
        
    except Exception as e:
        return False, f"Invalid URL format: {str(e)}"


def escape_sql_like(value: str) -> str:
    """
    Escape special characters in SQL LIKE patterns.
    
    Args:
        value: String to escape
        
    Returns:
        Escaped string safe for SQL LIKE clauses
    """
    if not value:
        return ""
    
    # Escape SQL LIKE wildcards
    value = value.replace('\\', '\\\\')
    value = value.replace('%', '\\%')
    value = value.replace('_', '\\_')
    
    return value


def sanitize_for_log(value: str, max_length: int = 200) -> str:
    """
    Sanitize string for safe logging (prevent log injection).
    
    Args:
        value: String to sanitize
        max_length: Maximum length for log output
        
    Returns:
        Sanitized string safe for logging
    """
    if not value:
        return ""
    
    # Remove newlines and carriage returns (log injection prevention)
    value = value.replace('\n', ' ').replace('\r', ' ')
    
    # Remove control characters
    value = ''.join(char for char in value if ord(char) >= 32 or char == ' ')
    
    # Truncate
    if len(value) > max_length:
        value = value[:max_length] + "..."
    
    return value


# Global validator instance
input_validator = InputValidator()


# Convenience functions
def validate_input(value: str, check_type: str = "comprehensive") -> Tuple[bool, Optional[str]]:
    """
    Convenience function for common validation.
    
    Args:
        value: String to validate
        check_type: Type of check ('sql', 'xss', 'path', 'command', 'comprehensive')
        
    Returns:
        Tuple of (is_safe, error_message)
    """
    if check_type == "sql":
        return input_validator.validate_sql_safe(value)
    elif check_type == "xss":
        return input_validator.validate_xss_safe(value)
    elif check_type == "path":
        return input_validator.validate_path_safe(value)
    elif check_type == "command":
        return input_validator.validate_command_safe(value)
    elif check_type == "comprehensive":
        is_safe, errors = input_validator.comprehensive_check(value)
        return is_safe, "; ".join(errors) if errors else None
    else:
        raise ValueError(f"Unknown check type: {check_type}")


__all__ = [
    'InputValidator',
    'sanitize_html',
    'sanitize_url',
    'validate_url',
    'escape_sql_like',
    'sanitize_for_log',
    'validate_input',
    'input_validator',
]

File: backend-ui/src/utils/validators.ts
/**
 * Frontend Input Validation & Sanitization Utilities
 * 
 * Provides client-side validation to complement server-side validation.
 * Never rely solely on client-side validation for security!
 * 
 * Usage:
 *   import { validateEmail, sanitizeString, isXSSSafe } from '@/utils/validators';
 *   
 *   if (!validateEmail(email)) {
 *     setError('Invalid email format');
 *   }
 */

// Dangerous patterns for XSS detection
const XSS_PATTERNS = [
  // Match script blocks even when closing tags contain whitespace/attributes
  // like </script\t\n bar>.
  /<script\b[^>]*>[\s\S]*?<\/script\b[^>]*>/gi,
  /javascript:/gi,
  /on\w+\s*=/gi,
  /<iframe[^>]*>/gi,
  /<embed[^>]*>/gi,
  /<object[^>]*>/gi,
  /eval\s*\(/gi,
  /expression\s*\(/gi,
  /vbscript:/gi,
  /data:text\/html/gi,
];

// SQL injection patterns
const SQL_PATTERNS = [
  /(\b(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE|UNION|SCRIPT)\b)/gi,
  /(--|#|\/\*|\*\/)/g,
  /(\bOR\b.*=.*)/gi,
  /(\bAND\b.*=.*)/gi,
];

// Path traversal patterns
const PATH_TRAVERSAL_PATTERNS = [
  /\.\.\//g,
  /\.\./g,
  /\/etc\//g,
  /\/proc\//g,
  /file:\/\//gi,
];

/**
 * Validation result type
 */
export interface ValidationResult {
  isValid: boolean;
  error?: string;
}

/**
 * Validate email format
 */
export function validateEmail(email: string): ValidationResult {
  if (!email) {
    return { isValid: false, error: 'Email cannot be empty' };
  }

  const emailPattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
  
  if (!emailPattern.test(email)) {
    return { isValid: false, error: 'Invalid email format' };
  }

  if (email.length > 254) {
    return { isValid: false, error: 'Email too long' };
  }

  return { isValid: true };
}

/**
 * Validate username format
 */
export function validateUsername(username: string): ValidationResult {
  if (!username) {
    return { isValid: false, error: 'Username cannot be empty' };
  }

  if (username.length < 3) {
    return { isValid: false, error: 'Username must be at least 3 characters' };
  }

  if (username.length > 50) {
    return { isValid: false, error: 'Username cannot exceed 50 characters' };
  }

  if (!/^[a-zA-Z0-9._-]+$/.test(username)) {
    return { isValid: false, error: 'Username can only contain letters, numbers, and ._- characters' };
  }

  return { isValid: true };
}

/**
 * Validate password strength
 */
export function validatePassword(password: string): ValidationResult {
  if (!password) {
    return { isValid: false, error: 'Password cannot be empty' };
  }

  if (password.length < 8) {
    return { isValid: false, error: 'Password must be at least 8 characters' };
  }

  if (!/[A-Z]/.test(password)) {
    return { isValid: false, error: 'Password must contain at least one uppercase letter' };
  }

  if (!/[a-z]/.test(password)) {
    return { isValid: false, error: 'Password must contain at least one lowercase letter' };
  }

  if (!/[0-9]/.test(password)) {
    return { isValid: false, error: 'Password must contain at least one number' };
  }

  if (!/[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]/.test(password)) {
    return { isValid: false, error: 'Password must contain at least one special character' };
  }

  return { isValid: true };
}

/**
 * Validate URL format
 */
export function validateUrl(url: string, allowedSchemes: string[] = ['http', 'https']): ValidationResult {
  if (!url) {
    return { isValid: false, error: 'URL cannot be empty' };
  }

  try {
    const parsed = new URL(url);
    
    if (!allowedSchemes.includes(parsed.protocol.replace(':', ''))) {
      return { isValid: false, error: `URL scheme must be one of: ${allowedSchemes.join(', ')}` };
    }

    if (url.length > 2048) {
      return { isValid: false, error: 'URL too long' };
    }

    return { isValid: true };
  } catch {
    return { isValid: false, error: 'Invalid URL format' };
  }
}

/**
 * Check if string contains XSS patterns
 */
export function isXSSSafe(value: string): ValidationResult {
  if (!value) {
    return { isValid: true };
  }

  for (const pattern of XSS_PATTERNS) {
    if (pattern.test(value)) {
      return { isValid: false, error: 'Potentially unsafe content detected' };
    }
  }

  return { isValid: true };
}

/**
 * Check if string contains SQL injection patterns
 */
export function isSQLSafe(value: string): ValidationResult {
  if (!value) {
    return { isValid: true };
  }

  for (const pattern of SQL_PATTERNS) {
    if (pattern.test(value)) {
      return { isValid: false, error: 'Potentially unsafe content detected' };
    }
  }

  return { isValid: true };
}

/**
 * Check if path contains traversal attempts
 */
export function isPathSafe(value: string): ValidationResult {
  if (!value) {
    return { isValid: true };
  }

  for (const pattern of PATH_TRAVERSAL_PATTERNS) {
    if (pattern.test(value)) {
      return { isValid: false, error: 'Invalid path format' };
    }
  }

  return { isValid: true };
}

/**
 * Sanitize string by removing dangerous characters
 */
export function sanitizeString(value: string, maxLength?: number): string {
  if (!value) {
    return '';
  }

  // Remove control characters (except common whitespace)
  let sanitized = value.replace(/[\x00-\x1F\x7F]/g, '');

  // Trim whitespace
  sanitized = sanitized.trim();

  // Truncate if needed
  if (maxLength && sanitized.length > maxLength) {
    sanitized = sanitized.substring(0, maxLength);
  }

  return sanitized;
}

/**
 * Sanitize HTML by escaping dangerous characters
 */
export function sanitizeHTML(html: string): string {
  if (!html) {
    return '';
  }

  const div = document.createElement('div');
  div.textContent = html;
  return div.innerHTML;
}

/**
 * Validate filename
 */
export function validateFilename(filename: string): ValidationResult {
  if (!filename) {
    return { isValid: false, error: 'Filename cannot be empty' };
  }

  if (filename.includes('..') || filename.includes('/') || filename.includes('\\')) {
    return { isValid: false, error: 'Filename cannot contain path separators' };
  }

  if (/[\x00-\x1F]/.test(filename)) {
    return { isValid: false, error: 'Filename contains invalid characters' };
  }

  if (filename.length > 255) {
    return { isValid: false, error: 'Filename too long' };
  }

  const dangerousExtensions = ['exe', 'bat', 'cmd', 'com', 'pif', 'scr', 'vbs', 'js', 'jar'];
  const ext = filename.split('.').pop()?.toLowerCase();
  
  if (ext && dangerousExtensions.includes(ext)) {
    return { isValid: false, error: `File extension '.${ext}' is not allowed` };
  }

  return { isValid: true };
}

/**
 * Validate slug format
 */
export function validateSlug(slug: string): ValidationResult {
  if (!slug) {
    return { isValid: false, error: 'Slug cannot be empty' };
  }

  if (!/^[a-z0-9-]+$/.test(slug)) {
    return { isValid: false, error: 'Slug must contain only lowercase letters, numbers, and hyphens' };
  }

  if (slug.startsWith('-') || slug.endsWith('-')) {
    return { isValid: false, error: 'Slug cannot start or end with a hyphen' };
  }

  if (slug.includes('--')) {
    return { isValid: false, error: 'Slug cannot contain consecutive hyphens' };
  }

  return { isValid: true };
}

/**
 * Validate phone number (basic international format)
 */
export function validatePhoneNumber(phone: string): ValidationResult {
  if (!phone) {
    return { isValid: false, error: 'Phone number cannot be empty' };
  }

  // Basic international phone format: +[country code][number]
  const phonePattern = /^\+?[1-9]\d{1,14}$/;
  
  // Remove common separators for validation
  const cleanPhone = phone.replace(/[\s\-\(\)]/g, '');

  if (!phonePattern.test(cleanPhone)) {
    return { isValid: false, error: 'Invalid phone number format' };
  }

  return { isValid: true };
}

/**
 * Validate hex color code
 */
export function validateHexColor(color: string): ValidationResult {
  if (!color) {
    return { isValid: false, error: 'Color cannot be empty' };
  }

  const hexPattern = /^#?([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/;

  if (!hexPattern.test(color)) {
    return { isValid: false, error: 'Invalid hex color format' };
  }

  return { isValid: true };
}

/**
 * Validate positive integer
 */
export function validatePositiveInteger(value: number, maxValue?: number): ValidationResult {
  if (!Number.isInteger(value)) {
    return { isValid: false, error: 'Value must be an integer' };
  }

  if (value < 1) {
    return { isValid: false, error: 'Value must be positive' };
  }

  if (maxValue && value > maxValue) {
    return { isValid: false, error: `Value cannot exceed ${maxValue}` };
  }

  return { isValid: true };
}

/**
 * Validate IP address (IPv4 or IPv6)
 */
export function validateIPAddress(ip: string): ValidationResult {
  if (!ip) {
    return { isValid: false, error: 'IP address cannot be empty' };
  }

  // IPv4 pattern
  const ipv4Pattern = /^(\d{1,3}\.){3}\d{1,3}$/;
  
  // IPv6 pattern (simplified)
  const ipv6Pattern = /^([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$/;

  if (ipv4Pattern.test(ip)) {
    // Validate each octet is 0-255
    const octets = ip.split('.');
    const validOctets = octets.every(octet => {
      const num = parseInt(octet, 10);
      return num >= 0 && num <= 255;
    });

    if (!validOctets) {
      return { isValid: false, error: 'Invalid IPv4 address' };
    }

    return { isValid: true };
  }

  if (ipv6Pattern.test(ip)) {
    return { isValid: true };
  }

  return { isValid: false, error: 'Invalid IP address format' };
}

/**
 * Comprehensive validation for form inputs
 */
export function validateInput(
  value: string,
  checks: Array<'xss' | 'sql' | 'path'> = ['xss', 'sql']
): ValidationResult {
  if (!value) {
    return { isValid: true };
  }

  if (checks.includes('xss')) {
    const xssResult = isXSSSafe(value);
    if (!xssResult.isValid) {
      return xssResult;
    }
  }

  if (checks.includes('sql')) {
    const sqlResult = isSQLSafe(value);
    if (!sqlResult.isValid) {
      return sqlResult;
    }
  }

  if (checks.includes('path')) {
    const pathResult = isPathSafe(value);
    if (!pathResult.isValid) {
      return pathResult;
    }
  }

  return { isValid: true };
}

/**
 * Create a string from maxLength that shows remaining characters
 */
export function getRemainingChars(current: string, maxLength: number): string {
  const remaining = maxLength - current.length;
  return remaining >= 0 ? `${remaining} characters remaining` : `${Math.abs(remaining)} characters over limit`;
}

/**
 * Validate credit card number using Luhn algorithm (for demo/testing purposes only)
 * DO NOT use for actual payment processing - use payment gateway APIs
 */
export function validateCreditCard(cardNumber: string): ValidationResult {
  if (!cardNumber) {
    return { isValid: false, error: 'Card number cannot be empty' };
  }

  // Remove spaces and dashes
  const cleanNumber = cardNumber.replace(/[\s\-]/g, '');

  // Check if only digits
  if (!/^\d+$/.test(cleanNumber)) {
    return { isValid: false, error: 'Card number must contain only digits' };
  }

  // Check length
  if (cleanNumber.length < 13 || cleanNumber.length > 19) {
    return { isValid: false, error: 'Invalid card number length' };
  }

  // Luhn algorithm
  let sum = 0;
  let isEven = false;

  for (let i = cleanNumber.length - 1; i >= 0; i--) {
    let digit = parseInt(cleanNumber[i], 10);

    if (isEven) {
      digit *= 2;
      if (digit > 9) {
        digit -= 9;
      }
    }

    sum += digit;
    isEven = !isEven;
  }

  if (sum % 10 !== 0) {
    return { isValid: false, error: 'Invalid card number' };
  }

  return { isValid: true };
}

const validators = {
  validateEmail,
  validateUsername,
  validatePassword,
  validateUrl,
  isXSSSafe,
  isSQLSafe,
  isPathSafe,
  sanitizeString,
  sanitizeHTML,
  validateFilename,
  validateSlug,
  validatePhoneNumber,
  validateHexColor,
  validatePositiveInteger,
  validateIPAddress,
  validateInput,
  getRemainingChars,
  validateCreditCard,
};

export default validators;
